### PR TITLE
Container refactor

### PR DESCRIPTION
**Pull request by stafyniaksacha**
# kuzzle-containers

This repository contains mutualized containers for running [Kuzzle](https://github.com/kuzzleio/kuzzle) stack in a debian environment.

It contains :
 * [kuzzleio/base](https://hub.docker.com/r/kuzzleio/base/) image: used to build kuzzleio projects in production mode
 * [kuzzleio/dev](https://hub.docker.com/r/kuzzleio/dev/) image: used to run kuzzleio projects in development mode

## kuzzleio/base

Contains nodejs (6.9.1) binary and pm2 global module

## kuzzleio/dev

Based on kuzzleio/base

Contains:
 * all needed to watch and build kuzzleio projects on the fly:
  * build-essential
  * g++
  * python
  * rbenv
  * libfontconfig
  * [gem] sass
  * [npm] bower
  * [npm] node-inspector
 * tools to improve development experience:
  * `ll` alias
  * bash-completion
  * curl
  * wget

# :warning: Associated PR:
* kuzzleio/kuzzle#421
* kuzzleio/kuzzle-proxy#24
* kuzzleio/kuzzle-backoffice#150
* kuzzleio/kuzzle-load-balancer#10
* kuzzleio/kuzzle-plugin-cluster#12

### CODE REVIEW

File: base/Dockerfile [hunk 1]
```diff
@@ -1,47 +1,15 @@
...
 
-WORKDIR /var/app
-CMD ["zsh"]
+RUN tar xf /tmp/node-v$NODE_VERSION-linux-x64.tar.gz -C /opt/ \
+    && rm /tmp/node-v$NODE_VERSION-linux-x64.tar.gz \
+    && mkdir -p /var/app \
+    && npm set progress=false \
+    && npm install -g \
+      pm2 \
+    && echo "" > /opt/node-v$NODE_VERSION-linux-x64/lib/node_modules/pm2/lib/keymetrics \
```
Comment (j33f) [inline base/Dockerfile]: maybe need to double check if the --silent options is not enough

File: README.md [hunk 2]
```diff
@@ -1,7 +1,32 @@
...
+  * build-essential
+  * g++
+  * python
+  * rbenv
+  * libfontconfig
+  * [gem] sass
+  * [npm] bower
+  * [npm] node-inspector
+ * tools to improve development experience:
+  * vim (+ profile)
```
Comment (ballinette) [inline README.md]: What's the aim of using vim (or any text editor) inside a container ?

Comment (stafyniaksacha) [inline README.md]: As the best practices said: there is no reason to have text editor or whatever in production mode.

But in development mode I think is quite good to have a text editor to test whatever you want

Comment (stafyniaksacha) [inline README.md]: As we said with the team: 
* removed vim & zsh
* added `ll` alias

File: README.md [hunk 3]
```diff
@@ -1,7 +1,31 @@
...
 
-# Documentation
+It contains :
+ * [kuzzleio/base](https://hub.docker.com/r/kuzzleio/base/) image: used to build kuzzleio projects in production mode
+ * [kuzzleio/dev](https://hub.docker.com/r/kuzzleio/dev/) image: used to run kuzzleio projects in development mode
 
-Please, refer to main project [Kuzzle](https://github.com/kuzzleio/kuzzle) for more information.
+## kuzzleio/base
+
+Contains nodejs (4.5.0) binary and pm2 global module
```
Comment (j33f) [inline README.md]: should be 4.6.0

File: base/README.md [hunk 4]
```diff
@@ -1,7 +1,5 @@
...
+# kuzzleio/base
 
-This repository contains Dockerfile for running [Kuzzle](https://github.com/kuzzleio/kuzzle).
+This repository contains Dockerfile for building [Kuzzle](https://github.com/kuzzleio/kuzzle) stack base image.
 
-# Documentation
-
-Please, refer to main project [Kuzzle](https://github.com/kuzzleio/kuzzle) for more information.
\ No newline at end of file
+Contains nodejs (4.5.0) binary and pm2 global module
```
Comment (j33f) [inline base/README.md]: ditto

Comment (stafyniaksacha) [inline base/README.md]: done

File: base/Dockerfile [hunk 5]
```diff
@@ -1,47 +1,15 @@
...
-
-RUN git clone https://github.com/gmarik/Vundle.vim.git ~/.vim/bundle/Vundle.vim && \
-    vim +PluginInstall +qall >/dev/null
-
-RUN mkdir -p /var/app
+ADD https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION-linux-x64.tar.gz /tmp/
 
-WORKDIR /var/app
-CMD ["zsh"]
+RUN tar xf /tmp/node-v$NODE_VERSION-linux-x64.tar.gz -C /opt/ \
```
Comment (benoitvidis) [inline base/Dockerfile]: As far as I can see, this dockerfile does the exact same thing as [the official Docker node image](https://github.com/nodejs/docker-node/blob/4029a8f71920e1e23efa79602167014f9c325ba0/4.6/Dockerfile) with the exception that the target installation directory for node is /opt here while Docker uses `/usr/local`

The later path has the advantage of automatically putting the binaries installed via `npm install -g` in the user path.

Is there any added-value I missed to manually install node by ourselves or can we just use Docker's official image as base?

File: dev/Dockerfile [hunk 6]
```diff
@@ -1,17 +1,27 @@
...
+RUN apt-get update && apt-get install -y \
+      build-essential \
+      g++ \
+      python \
+      libfontconfig \
+      bash-completion \
+      rbenv \
+      curl \
+      git \
+      wget \
```
Comment (benoitvidis) [inline dev/Dockerfile]: nitpicking: can we sort the list of packages by alphabetical order?
+ curl is already installed on node:* official images.

File: base/README.md [hunk 7]
```diff
@@ -1,7 +1,5 @@
...
+# kuzzleio/base
 
-This repository contains Dockerfile for running [Kuzzle](https://github.com/kuzzleio/kuzzle).
+This repository contains Dockerfile for building [Kuzzle](https://github.com/kuzzleio/kuzzle) stack base image.
 
-# Documentation
-
-Please, refer to main project [Kuzzle](https://github.com/kuzzleio/kuzzle) for more information.
\ No newline at end of file
+Contains nodejs (4.6.0) binary and pm2 global module
```
Comment (benoitvidis) [inline base/README.md]: To me, this file does not bring any added value. I would remove it.

File: dev/Dockerfile (hunk 6 above)
Comment (benoitvidis) [inline dev/Dockerfile]: is wget usefull if we already have curl?

File: dev/README.md [hunk 8]
```diff
@@ -1,7 +1,19 @@
-# Kuzzle
```
Comment (benoitvidis) [inline dev/README.md]: Same as for base/README.md, this is just a duplicate of the main readme, without any added value.
I would delete this file.

File: dev/profile/.vimrc [hunk 9]
```diff
@@ -14,7 +14,7 @@ filetype plugin indent on
 
 syntax on
 set background=dark
-color hybrid
+color codeschool
 set laststatus=2
 set expandtab
 set ts=4
```
Comment (benoitvidis) [inline dev/profile/.vimrc]: ts & sw should most likely be set to 2 imo

File: base/Dockerfile (hunk 1 above)
Comment (xbill82) [inline base/Dockerfile]: https://github.com/npm/npm/issues/2040

File: base/README.md (hunk 7 above)
Comment (stafyniaksacha) [inline base/README.md]: This file is only displayed on the docker hub.

Did we need to remove it ?

File: dev/README.md (hunk 8 above)
Comment (stafyniaksacha) [inline dev/README.md]: This file is only displayed on the docker hub.

Did we need to remove it ?

File: base/Dockerfile (hunk 5 above)
Comment (stafyniaksacha) [inline base/Dockerfile]: The official node image is builded on [buildpack-deps](https://hub.docker.com/_/buildpack-deps/) image which take 500 mo more than the [debian](https://hub.docker.com/_/debian/) image.

In top of that it's embed a lot of dev packages, such as imagemagic, libmysqlclient, ....

Comment (benoitvidis) [inline base/Dockerfile]: you're right. I had missed the fact that we we talking about the base image. 
LGTM

File: base/README.md (hunk 7 above)
Comment (benoitvidis) [inline base/README.md]: ah, true. I had missed he hub.
:+1:

File: base/Dockerfile [hunk 10]
```diff
@@ -1,47 +1,15 @@
...
-    zsh && \
-    apt-get clean
-
-ENV NODE_VERSION=4.2.1
-
-ADD https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION-linux-x64.tar.gz /root/
-RUN tar xvf /root/node-v$NODE_VERSION-linux-x64.tar.gz -C /opt/ && \
-    rm /root/node-v$NODE_VERSION-linux-x64.tar.gz
-
+ENV NODE_VERSION=4.6.0
```
Comment (dbengsch) [inline base/Dockerfile]: What about upgrading to 6.9.1 LTS

**Pull request merged**